lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -50,15 +50,29 @@ def initialize(serializer, options = {})
         def serializable_hash(options = nil)
           options ||= {}
 
-          hash =
-            if serializer.respond_to?(:each)
-              serializable_hash_for_collection(options)
+          is_collection = serializer.respond_to?(:each)
+          resources = {}
+          process_resource(serializer, @include_tree, true, resources)
+          primary_data, included = resources.each_with_object([[], []]) do |(_, resource), (primary, included)|
+            if resource[:is_primary]
+              primary.push(resource.except(:is_primary))
             else
-              serializable_hash_for_single_resource
+              included.push(resource.except(:is_primary))
             end
+          end
```

Comment:
This block could also be rewritten in the following shorter form:
```ruby
primary_data, included = resources.partition { |_, r| r[:is_primary] }.map { |s| s.map { |_, r| r.except(:is_primary) } }  
```
which I am sure @bf4 would love haha